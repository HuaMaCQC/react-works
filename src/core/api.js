import Axios from 'axios';

const config = require('../config.json');

const ApiRequest = Axios.create({
  baseURL: config.ApiHost,
});

export const apiGetUser = data => ApiRequest.get('api/users', { params: data });
export const apiAddUser = data => ApiRequest.post('api/user', data);
export const apiDeleteUser = id => ApiRequest.delete(`api/user/${id}`);
export const apiEditUser = (data, id) => ApiRequest.put(`api/user/${id}`, data);
