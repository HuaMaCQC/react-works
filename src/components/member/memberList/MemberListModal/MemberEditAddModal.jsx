import React, { useState } from 'react';
import {
  Button, Modal, Form, Label,
} from 'semantic-ui-react';
import { apiAddUser, apiEditUser } from '../../../../core/api';

export default function MemberEditAddModal(props) {
  const {
    modalData,
    open,
    close,
    onComplete,
  } = props;
  const [username, setUsername] = useState(modalData.username);
  const [enable, setEnable] = useState(modalData.enable);
  const [locked, setLocked] = useState(modalData.locked);

  const type = {
    insert: 'insert',
    edit: 'edit',
  };
  const enableOptions = [
    { key: 1, text: '啟用', value: 1 },
    { key: 0, text: '未啟用', value: 0 },
  ];
  const lockedOptions = [
    { key: 1, text: '鎖定', value: 1 },
    { key: 0, text: '未鎖定', value: 0 },
  ];

  function save() {
    if (modalData.type === type.insert) {
      const addData = {
        username,
        enable,
        locked,
      };

      apiAddUser(addData).then(() => {
        onComplete();
      });
    }

    if (modalData.type === type.edit) {
      const editData = {
        username,
        enable,
        locked,
      };

      apiEditUser(editData, modalData.id).then(() => {
        onComplete();
      });
    }
  }

  return (
    <div>
      <Modal size="large" open={open} onClose={close}>
        <Modal.Header data-testid="title" content={modalData.title} />
        <Modal.Content>
          <Form data-testid="form">
            {
              !!modalData.id && (
                <Label.Group size="huge">
                  <Label content={`id : ${modalData.id}`} />
                </Label.Group>
              )
            }
            <Form.Input
              fluid
              label="名子"
              name="username"
              placeholder="請輸入姓名"
              data-testid="username"
              defaultValue={modalData.username}
              onChange={(e, data) => setUsername(data.value)}
            />
            <Form.Select
              fluid
              name="enable"
              label="狀態"
              data-testid="enable"
              options={enableOptions}
              defaultValue={modalData.enable}
              onChange={(e, data) => setEnable(data.value)}
            />
            <Form.Select
              fluid
              name="locked"
              label="鎖定"
              data-testid="locked"
              options={lockedOptions}
              defaultValue={modalData.locked}
              onChange={(e, data) => setLocked(data.value)}
            />
          </Form>
        </Modal.Content>
        <Modal.Actions>
          <Button data-testid="close" color="red" content="取消" onClick={close} />
          <Button data-testid="save" content="完成" onClick={save} primary />
        </Modal.Actions>
      </Modal>
    </div>
  );
}
