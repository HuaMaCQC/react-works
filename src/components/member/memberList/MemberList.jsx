import React, { useState, useEffect, useReducer } from 'react';
import {
  Table, Pagination, Grid, Form, Icon, Button,
} from 'semantic-ui-react';
import { DateTimeInput } from 'semantic-ui-calendar-react';
import moment from 'moment';
import styles from './MemberList.styledJsx.scss';
import { apiGetUser, apiDeleteUser } from '../../../core/api';
import MemberEditAddModal from './MemberListModal/MemberEditAddModal';
import createCsvFile from '../../../public/CreateCsvFile';

const initfindData = {
  username: undefined,
  enable: undefined,
  locked: undefined,
  start_created_at: undefined,
  end_created_at: undefined,
  max_results: 5,
  first_result: 0,
};

function handleFindChange(state, { findData, name, value }) {
  let v = value;

  if (name === 'start_created_at' || name === 'end_created_at') {
    v = value === '' ? '' : moment(value).format();
  }

  return { ...findData || state, [name]: v === '' ? undefined : v };
}

export default function MemberList() {
  const [memberData, setMemberData] = useState([]);
  const [findData, setFindData] = useReducer(handleFindChange, initfindData);
  const [csv, setCsv] = useState({ url: '', fileName: '' });
  const [pagination, setPagination] = useState(
    {
      nowActivePage: 1,
      totalPages: 0,
      total: 0,
    },
  );
  const [modalData, setModalData] = useState(
    {
      id: undefined,
      username: undefined,
      enable: 1,
      locked: 0,
      title: '',
      type: '',
    },
  );
  const [open, setOpen] = useState(false);
  const [startCreatedAt, setStartCreatedAt] = useState(undefined);
  const [endCreatedAt, setEndCreatedAt] = useState(undefined);
  const type = { insert: 'insert', edit: 'edit' };
  const {
    nowActivePage,
    totalPages,
  } = pagination;
  const enableOptions = [
    { key: '-1', text: '全部', value: undefined },
    { key: '1', text: '啟用', value: 1 },
    { key: '0', text: '未啟用', value: 0 },
  ];
  const lockedOptions = [
    { key: '-1', text: '全部', value: undefined },
    { key: '1', text: '鎖定', value: 1 },
    { key: '0', text: '未鎖定', value: 0 },
  ];
  const jsxId = (() => {
    const jsx = <scope><style jsx>{styles}</style></scope>;
    return jsx.props.className;
  })();

  function getUser(newFindData) {
    apiGetUser(newFindData).then((res) => {
      const maxResults = Number(res.data.pagination.max_results);
      const total = Number(res.data.pagination.total);
      const newActivePage = Math.ceil(Number(res.data.pagination.first_result) / maxResults + 1);

      setMemberData(res.data.ret);
      setPagination(
        {
          total,
          nowActivePage: newActivePage,
          totalPages: Math.ceil(total / maxResults),
        },
      );
    });
  }

  // eslint-disable-next-line
  useEffect(() => { getUser(findData); }, []);

  function search() {
    const nextFindData = {
      ...findData,
      first_result: 0,
    };

    setFindData({ findData: nextFindData });
    getUser(nextFindData);
  }

  function handlePaginationChange({ activePage }) {
    const nextFindData = {
      ...findData,
      first_result: (activePage - 1) * findData.max_results,
    };

    setFindData({ findData: nextFindData });
    getUser(nextFindData);
  }

  function openModal(opentype, data) {
    if (opentype === type.insert) {
      setModalData({
        type: type.insert,
        id: 0,
        username: '',
        enable: 1,
        locked: 0,
        title: '新增',
      });
    }

    if (opentype === type.edit) {
      setModalData({
        type: type.edit,
        id: data.id,
        username: data.username || '',
        enable: data.enable === undefined ? 0 : data.enable,
        locked: data.locked === undefined ? 1 : data.locked,
        title: '修改',
      });
    }

    setOpen(true);
  }

  function close() {
    setOpen(false);
  }

  function onComplete() {
    getUser(findData);
    close();
  }

  function del(e) {
    const { id } = e.currentTarget.dataset;
    apiDeleteUser(id).then(() => {
      getUser(findData);
    });
  }

  function changeSearchStartTime(e, data) {
    setStartCreatedAt(data.value);
    setFindData(data);
  }

  function changeSearchEndTime(e, data) {
    setEndCreatedAt(data.value);
    setFindData(data);
  }

  function exportCsv() {
    const data = memberData.reduce((accumulator, value) => {
      accumulator.push([
        value.id,
        value.username,
        value.enable ? '啟動' : '未啟動',
        value.locked ? '鎖定' : '未鎖定',
        moment(value.created_at).format('YYYY-MM-DD hh:mm:ss'),
      ]);

      return accumulator;
    }, [['id', '姓名', '狀態', '鎖定', '創建時間']]);

    const myCsv = createCsvFile(data);

    setCsv({ url: myCsv, fileName: '會員資料.csv' });
  }

  return (
    <div className="member-list">
      <Grid columns={1}>
        <Grid.Column className={`${jsxId} grid-card`}>
          <Form>
            <Form.Group widths="equal">
              <Form.Input
                fluid
                label="姓名"
                placeholder="請輸入姓名"
                name="username"
                data-testid="search-username"
                onChange={(e, data) => setFindData(data)}
              />
              <Form.Select
                fluid
                name="enable"
                label="狀態"
                data-testid="search-enable"
                options={enableOptions}
                placeholder={enableOptions[0].text}
                onChange={(e, data) => setFindData(data)}
              />
              <Form.Select
                fluid
                name="locked"
                label="鎖定"
                data-testid="search-locked"
                options={lockedOptions}
                placeholder={enableOptions[0].text}
                onChange={(e, data) => setFindData(data)}
              />
            </Form.Group>
            <Form.Group inline>
              查詢時間
              <DateTimeInput
                clearable
                closable
                name="start_created_at"
                animation="none"
                placeholder="全部時間"
                popupPosition="bottom left"
                dateTimeFormat="YYYY-MM-DD hh:mm:ss"
                data-testid="search-start_created_at"
                value={startCreatedAt}
                clearIcon={<Icon name="remove" />}
                onChange={changeSearchStartTime}
              />
              ~
              <DateTimeInput
                clearable
                closable
                name="end_created_at"
                animation="none"
                placeholder="全部時間"
                dateTimeFormat="YYYY-MM-DD hh:mm:ss"
                data-testid="search-end_created_at"
                popupPosition="bottom left"
                value={endCreatedAt}
                clearIcon={<Icon name="remove" />}
                onChange={changeSearchEndTime}
              />
            </Form.Group>
            <Form.Group className={`${jsxId} float-right`}>
              <Form.Button onClick={search} data-testid="search-btn" content="查詢" />
            </Form.Group>
          </Form>
        </Grid.Column>
        <Grid.Column className={`${jsxId} grid-content`}>
          <Table celled>
            <Table.Header data-testid="table-header">
              <Table.Row>
                <Table.HeaderCell content="ID" />
                <Table.HeaderCell content="姓名" />
                <Table.HeaderCell content="狀態" />
                <Table.HeaderCell content="鎖定" />
                <Table.HeaderCell content="創建時間" />
                <Table.HeaderCell content="操作" />
              </Table.Row>
            </Table.Header>
            <Table.Body data-testid="table-list">
              {
                memberData.map(data => (
                  <Table.Row key={data.id}>
                    <Table.Cell content={data.id} />
                    <Table.Cell content={data.username} />
                    <Table.Cell
                      className={`${jsxId} ${data.enable ? 'member-not-lock' : 'member-lock'}`}
                      content={data.enable ? '啟動' : '未啟動'}
                    />
                    <Table.Cell className={`${jsxId} ${data.locked ? 'member-lock' : 'member-not-lock'}`}>
                      <Icon name={data.locked ? 'lock' : 'lock open'} />
                      {data.locked ? '鎖定' : '未鎖定'}
                    </Table.Cell>
                    <Table.Cell content={moment(data.created_at).format('YYYY-MM-DD hh:mm:ss')} />
                    <Table.Cell>
                      <Button
                        content="編輯"
                        onClick={() => openModal(type.edit, data)}
                      />
                      <Button
                        content="刪除"
                        data-id={data.id}
                        onClick={del}
                      />
                    </Table.Cell>
                  </Table.Row>
                ))
              }
            </Table.Body>
            <Table.Footer>
              <Table.Row>
                <Table.HeaderCell colSpan="5">
                  <Pagination
                    size="mini"
                    siblingRange="2"
                    data-testid="table-pagination"
                    activePage={nowActivePage}
                    totalPages={totalPages}
                    onPageChange={(e, data) => handlePaginationChange(data)}
                  />
                </Table.HeaderCell>
                <Table.HeaderCell colSpan="1">
                  <Button data-testid="member-add" onClick={() => openModal(type.insert)} content="新增" />
                  <Button
                    as="a"
                    data-testid="export-csv"
                    content="匯出"
                    href={csv.url}
                    download={csv.fileName}
                    onClick={exportCsv}
                  />
                </Table.HeaderCell>
              </Table.Row>
            </Table.Footer>
          </Table>
        </Grid.Column>
      </Grid>
      <MemberEditAddModal
        modalData={modalData}
        close={close}
        open={open}
        onComplete={onComplete}
      />
      <style jsx="true">{styles}</style>
    </div>
  );
}
