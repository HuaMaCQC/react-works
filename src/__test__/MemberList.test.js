
/* eslint-disable react/jsx-filename-extension */
/* eslint-disable no-undef */

import React from 'react';
import {
  render,
  fireEvent,
  cleanup,
  waitForElement,
} from '@testing-library/react';
import 'jest-dom/extend-expect';
// 要測試的組件
import MemberList from '../components/member/memberList/MemberList';
import { apiGetUser, apiDeleteUser } from '../core/api';
import MemberEditAddModal from '../components/member/memberList/MemberListModal/MemberEditAddModal';
import createCsvFile from '../public/CreateCsvFile';

const moment = require.requireActual('moment-timezone');
const apiMock = require('../__mocks__/ApiMemberListMocks');

const originalError = console.error;

jest.mock('../components/member/memberList/MemberListModal/MemberEditAddModal');
jest.mock('../core/api');
jest.mock('../public/CreateCsvFile');
moment.tz.setDefault('Asia/Taipei');

beforeAll(() => {
  console.error = jest.fn();
});

afterAll(() => {
  console.error = originalError;
});

describe('search UI', () => {
  let component = {
    getByTestId: null,
  };

  beforeAll(() => {
    cleanup();
    MemberEditAddModal.mockImplementation(jest.fn(() => <div />));
    apiGetUser.mockImplementation(apiMock.apiGetUser);
    component = render(<MemberList />);
  });

  it('search-username ', async () => {
    const { getByTestId } = component;
    const username = getByTestId('search-username').querySelector('input');

    fireEvent.change(username, { target: { value: 'test' } });

    expect(getByTestId('search-username')).toBeTruthy();
    expect(username.placeholder).toBe('請輸入姓名');
    expect(username.value).toBe('test');
  });

  it('search-enable ', () => {
    const { getByTestId } = component;
    const enableOption = getByTestId('search-enable').querySelectorAll('span');
    const enableBtn = getByTestId('search-enable').querySelectorAll('div')[3];
    const value = getByTestId('search-enable').querySelectorAll('div')[0];

    expect(value).toHaveTextContent(/^全部$/);

    fireEvent.click(enableBtn);

    expect(getByTestId('search-enable')).toBeTruthy();
    expect(enableOption).toHaveLength(3);
    expect(value).toHaveTextContent(/^啟用$/);
  });

  it('search-locked ', () => {
    const { getByTestId } = component;
    const lockedOption = getByTestId('search-locked').querySelectorAll('span');
    const lockedBtn = getByTestId('search-locked').querySelectorAll('div')[3];
    const value = getByTestId('search-locked').querySelectorAll('div')[0];

    expect(value).toHaveTextContent(/^全部$/);

    fireEvent.click(lockedBtn);

    expect(value).toHaveTextContent(/^鎖定$/);
    expect(getByTestId('search-locked')).toBeTruthy();
    expect(lockedOption).toHaveLength(3);
  });

  it('search-start_created_at ', () => {
    const { getByTestId } = component;
    const startCreatedAtInput = getByTestId('search-start_created_at').querySelector('input');

    fireEvent.change(startCreatedAtInput, { target: { value: '2000-01-01 00:00:00' } });

    expect(getByTestId('search-start_created_at')).toBeTruthy();
    expect(startCreatedAtInput.placeholder).toBe('全部時間');
    expect(startCreatedAtInput.value).toBe('2000-01-01 00:00:00');
  });

  it('search-end_created_at ', () => {
    const { getByTestId } = component;
    const endCreatedAtInput = getByTestId('search-end_created_at').querySelector('input');

    fireEvent.change(endCreatedAtInput, { target: { value: '2000-01-01 00:00:00' } });

    expect(getByTestId('search-end_created_at')).toBeTruthy();
    expect(endCreatedAtInput.placeholder).toBe('全部時間');
    expect(endCreatedAtInput.value).toBe('2000-01-01 00:00:00');

    fireEvent.change(endCreatedAtInput, { target: { value: '' } });
    expect(endCreatedAtInput.value).toBe('');
  });

  it('search button', () => {
    const { getByTestId } = component;
    const mockResData = {
      data: {
        result: 'ok',
        ret: [
          {
            created_at: '2019-01-16T21:36:04+08:00',
            enable: 1,
            id: 1,
            locked: 1,
            username: 'user1',
          },
        ],
        pagination: {
          first_result: '0',
          max_results: '5',
          total: 50,
        },
      },
    };
    const mockFindData = {
      username: 'test',
      enable: 1,
      locked: 1,
      start_created_at: moment('2000-01-01 00:00:00').format(),
      end_created_at: undefined,
      max_results: 5,
      first_result: 0,
    };
    const apiGetUserFn = jest.fn(() => Promise.resolve(mockResData));

    apiGetUser.mockImplementation(apiGetUserFn);

    const button = getByTestId('search-btn');

    fireEvent.click(button);

    expect(getByTestId('search-btn')).toBeTruthy();
    expect(button).toHaveTextContent(/^查詢$/);
    expect(apiGetUserFn).toHaveBeenCalledTimes(1);
    expect(apiGetUserFn).toHaveBeenCalledWith(mockFindData);
  });
});

describe('Member Table', () => {
  beforeAll(() => {
    cleanup();
  });

  beforeEach(() => {
    MemberEditAddModal.mockImplementation(jest.fn(() => <div />));
    apiGetUser.mockImplementation(apiMock.apiGetUser);
  });

  afterEach(() => {
    cleanup();
    jest.clearAllMocks();
  });

  it('table header', () => {
    const { getByTestId } = render(<MemberList />);
    const header = getByTestId('table-header').querySelectorAll('th');

    expect(header).toHaveLength(6);
  });

  it('table 資料', async () => {
    const { getByTestId } = render(<MemberList />);
    const table = await waitForElement(() => getByTestId('table-list'));
    const tds0 = table.querySelectorAll('tr')[0].querySelectorAll('td');
    const tds1 = table.querySelectorAll('tr')[1].querySelectorAll('td');
    const time = moment('2019-06-16T21:36:04+08:00').format('YYYY-MM-DD hh:mm:ss');

    expect(table.querySelectorAll('tr')).toHaveLength(5);
    expect(tds0[0]).toHaveTextContent(/^1$/);
    expect(tds0[1]).toHaveTextContent(/^user1$/);
    expect(tds0[2]).toHaveTextContent(/^啟動$/);
    expect(tds0[3]).toHaveTextContent(/^鎖定$/);
    expect(tds0[4]).toHaveTextContent(time);
    expect(tds0[2]).toHaveClass('member-not-lock');
    expect(tds0[3]).toHaveClass('member-lock');
    expect(tds1[2]).toHaveTextContent(/^未啟動$/);
    expect(tds1[3]).toHaveTextContent(/^未鎖定$/);
    expect(tds1[2]).toHaveClass('member-lock');
    expect(tds1[3]).toHaveClass('member-not-lock');
  });

  it('table Snapshot Testing', async () => {
    const { getByTestId } = render(<MemberList />);
    const table = await waitForElement(() => getByTestId('table-list'));

    expect(table).toMatchSnapshot();
  });

  it('table delete button', async () => {
    // 初始化模擬資料
    const initData = {
      data: {
        result: 'ok',
        ret: [
          {
            created_at: '2019-01-16T21:36:04+08:00',
            enable: 1,
            id: 298,
            locked: 1,
            username: 'user298',
          },
          {
            created_at: '2019-01-16T21:36:04+08:00',
            enable: 1,
            id: 299,
            locked: 1,
            username: 'user299',
          },
        ],
        pagination: {
          first_result: '0',
          max_results: '5',
          total: 50,
        },
      },
    };
    const apiGetUserFnInit = jest.fn(() => Promise.resolve(initData));
    const apiDeleteUserFn = jest.fn(() => Promise.resolve());
    apiDeleteUser.mockImplementation(apiDeleteUserFn);

    apiGetUser.mockImplementation(apiGetUserFnInit);

    const { getByTestId, rerender } = render(<MemberList />);

    const table = await waitForElement(() => getByTestId('table-list'));
    const delBtn = table
      .querySelector('tr')
      .querySelectorAll('td')[5]
      .querySelectorAll('button')[1];

    fireEvent.click(delBtn);

    // 模擬刪除後回傳
    const delFinishData = {
      data: {
        result: 'ok',
        ret: [
          {
            created_at: '2019-01-16T21:36:04+08:00',
            enable: 1,
            id: 299,
            locked: 1,
            username: 'user299',
          },
        ],
        pagination: {
          first_result: '0',
          max_results: '5',
          total: 50,
        },
      },
    };
    const apiGetUserFn = jest.fn(() => Promise.resolve(delFinishData));
    apiGetUser.mockImplementation(apiGetUserFn);

    rerender(<MemberList />);

    const delFinshTable = await waitForElement(() => getByTestId('table-list'));
    const id = delFinshTable
      .querySelector('tr')
      .querySelectorAll('td')[0];

    expect(apiDeleteUserFn).toHaveBeenCalledTimes(1);
    expect(id).toHaveTextContent(/^299$/);
    expect(apiDeleteUserFn).toHaveBeenCalledWith('298');
    expect(delBtn).toHaveTextContent(/^刪除$/);
  });

  it('table Pagination', async () => {
    const { getByTestId, rerender } = render(<MemberList />);
    const pagination = await waitForElement(() => getByTestId('table-pagination'));
    const pageBen = pagination.querySelectorAll('a');

    const nextPageData = {
      data: {
        result: 'ok',
        ret: [
          {
            created_at: '2019-01-16T21:36:04+08:00',
            enable: 1,
            id: 666,
            locked: 1,
            username: 'user666',
          },
        ],
        pagination: {
          first_result: '5',
          max_results: '1',
          total: 50,
        },
      },
    };
    const apiGetUserNextPageData = jest.fn(() => Promise.resolve(nextPageData));
    apiGetUser.mockImplementation(apiGetUserNextPageData);

    fireEvent.click(pageBen[3]);


    rerender(<MemberList />);

    const table = await waitForElement(() => getByTestId('table-list'));
    const id = table.querySelector('tr').querySelectorAll('td')[0];
    const findData = {
      enable: undefined,
      end_created_at: undefined,
      first_result: 5,
      locked: undefined,
      max_results: 5,
      start_created_at: undefined,
      username: undefined,
    };

    expect(apiGetUserNextPageData).toHaveBeenCalledWith(findData);
    expect(apiGetUserNextPageData).toHaveBeenCalledTimes(1);
    expect(id).toHaveTextContent(/^666$/);
    expect(pageBen[3]).toHaveTextContent(/^2$/);
    expect(pageBen).toHaveLength(13);
  });

  it('member Add', async () => {
    let propsData = {};
    const mockElement = (data) => {
      propsData = data;
      return (<div />);
    };
    MemberEditAddModal.mockImplementation(mockElement);
    const { getByTestId, rerender } = render(<MemberList />);
    const addBtn = await waitForElement(() => getByTestId('member-add'));
    // 按下新增
    fireEvent.click(addBtn);

    expect(addBtn).toHaveTextContent(/^新增$/);
    expect(propsData.open).toBe(true);
    expect(propsData.modalData).toEqual({
      type: 'insert',
      id: 0,
      username: '',
      enable: 1,
      locked: 0,
      title: '新增',
    });

    // 執行關閉
    propsData.close();
    expect(propsData.open).toBe(false);

    const delFinishData = {
      data: {
        result: 'ok',
        ret: [
          {
            created_at: '2019-01-16T21:36:04+08:00',
            enable: 1,
            id: 8,
            locked: 1,
            username: 'user299',
          },
        ],
        pagination: {
          first_result: '0',
          max_results: '5',
          total: 50,
        },
      },
    };
    const apiGetUserFn = jest.fn(() => Promise.resolve(delFinishData));
    apiGetUser.mockImplementation(apiGetUserFn);

    // 按下新增
    fireEvent.click(addBtn);
    // 執行完成
    propsData.onComplete();

    rerender(<MemberList />);

    const table = await waitForElement(() => getByTestId('table-list'));
    const id = table.querySelector('tr').querySelectorAll('td')[0];

    expect(propsData.open).toBe(false);
    expect(apiGetUserFn).toHaveBeenCalledTimes(1);
    expect(id).toHaveTextContent(/^8$/);
  });

  it('export Csv', async () => {
    // 模擬apiGetUser
    const mockApiGetUserData = {
      data: {
        result: 'ok',
        ret: [
          {
            created_at: '2019-01-16T21:36:04+08:00',
            enable: 1,
            id: 88,
            locked: 1,
            username: 'isCsvTest',
          },
        ],
        pagination: {
          first_result: '0',
          max_results: '5',
          total: 50,
        },
      },
    };
    apiGetUser.mockImplementation(jest.fn(() => Promise.resolve(mockApiGetUserData)));

    // 模擬創建下載點
    const mockCreateCsvFile = jest.fn(() => '/download-csv');
    createCsvFile.mockImplementation(mockCreateCsvFile);

    // 開始測試
    const { getByTestId } = render(<MemberList />);
    const exportCsvBtn = await waitForElement(() => getByTestId('export-csv'));

    fireEvent.click(exportCsvBtn);

    const newBtn = await waitForElement(() => getByTestId('export-csv'));

    expect(newBtn.download).toBe('會員資料.csv');
    expect(mockCreateCsvFile).toHaveBeenCalledTimes(1);
    expect(mockCreateCsvFile).toHaveBeenCalledWith(
      [
        ['id', '姓名', '狀態', '鎖定', '創建時間'],
        [88, 'isCsvTest', '啟動', '鎖定', '2019-01-16 09:36:04'],
      ],
    );
    expect(newBtn.href).toBe('http://localhost/download-csv');

    // 模擬二次取得資料
    const mockApiGetUserData2 = {
      data: {
        result: 'ok',
        ret: [
          {
            created_at: '2019-01-16T21:36:04+08:00',
            enable: 0,
            id: 66,
            locked: 0,
            username: 'isCsvTest2',
          },
        ],
        pagination: {
          first_result: '0',
          max_results: '5',
          total: 50,
        },
      },
    };
    apiGetUser.mockImplementation(jest.fn(() => Promise.resolve(mockApiGetUserData2)));

    const searchBtn = getByTestId('search-btn');

    await fireEvent.click(searchBtn);
    fireEvent.click(exportCsvBtn);

    expect(mockCreateCsvFile).toHaveBeenCalledTimes(2);
    expect(mockCreateCsvFile).toHaveBeenLastCalledWith(
      [
        ['id', '姓名', '狀態', '鎖定', '創建時間'],
        [66, 'isCsvTest2', '未啟動', '未鎖定', '2019-01-16 09:36:04'],
      ],
    );
  });

  it('member Edit', async () => {
    const mockApiGetUserData = {
      data: {
        result: 'ok',
        ret: [
          {
            created_at: '2019-01-16T21:36:04+08:00',
            enable: undefined,
            id: 9,
            locked: undefined,
            username: undefined,
          },
        ],
        pagination: {
          first_result: '0',
          max_results: '5',
          total: 50,
        },
      },
    };
    apiGetUser.mockImplementation(jest.fn(() => Promise.resolve(mockApiGetUserData)));

    let propsData = {};
    const mockElement = (data) => {
      propsData = data;
      return (<div />);
    };
    MemberEditAddModal.mockImplementation(mockElement);

    const { getByTestId, rerender } = render(<MemberList />);
    const table = await waitForElement(() => getByTestId('table-list'));
    const editBtn = table
      .querySelector('tr')
      .querySelectorAll('td')[5]
      .querySelectorAll('button')[0];

    // // 按下修改
    fireEvent.click(editBtn);

    expect(editBtn).toHaveTextContent(/^編輯$/);
    expect(propsData.open).toBe(true);
    expect(propsData.modalData).toEqual({
      type: 'edit',
      id: 9,
      username: '',
      enable: 0,
      locked: 1,
      title: '修改',
    });

    const data2 = {
      data: {
        result: 'ok',
        ret: [
          {
            created_at: '2019-01-16T21:36:04+08:00',
            enable: 1,
            id: 10,
            locked: 1,
            username: 'userEdit',
          },
        ],
        pagination: {
          first_result: '0',
          max_results: '5',
          total: 50,
        },
      },
    };
    apiGetUser.mockImplementation(jest.fn(() => Promise.resolve(data2)));

    // 完成關閉
    propsData.onComplete();

    rerender(<MemberList />);

    const table2 = await waitForElement(() => getByTestId('table-list'));
    const editBtn2 = table2
      .querySelector('tr')
      .querySelectorAll('td')[5]
      .querySelectorAll('button')[0];

    // 按下修改
    fireEvent.click(editBtn2);

    expect(propsData.modalData).toEqual(
      {
        type: 'edit',
        id: 10,
        username: 'userEdit',
        enable: 1,
        locked: 1,
        title: '修改',
      },
    );
  });
});
