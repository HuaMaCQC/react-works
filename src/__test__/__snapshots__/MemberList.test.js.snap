// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Member Table table Snapshot Testing 1`] = `
<tbody
  class=""
  data-testid="table-list"
>
  <tr
    class=""
  >
    <td
      class=""
    >
      1
    </td>
    <td
      class=""
    >
      user1
    </td>
    <td
      class="undefined member-not-lock"
    >
      啟動
    </td>
    <td
      class="undefined member-lock"
    >
      <i
        aria-hidden="true"
        class="lock icon"
      />
      鎖定
    </td>
    <td
      class=""
    >
      2019-06-16 09:36:04
    </td>
    <td
      class=""
    >
      <button
        class="ui button"
      >
        編輯
      </button>
      <button
        class="ui button"
        data-id="1"
      >
        刪除
      </button>
    </td>
  </tr>
  <tr
    class=""
  >
    <td
      class=""
    >
      2
    </td>
    <td
      class=""
    >
      user2
    </td>
    <td
      class="undefined member-lock"
    >
      未啟動
    </td>
    <td
      class="undefined member-not-lock"
    >
      <i
        aria-hidden="true"
        class="lock open icon"
      />
      未鎖定
    </td>
    <td
      class=""
    >
      2019-06-16 09:36:04
    </td>
    <td
      class=""
    >
      <button
        class="ui button"
      >
        編輯
      </button>
      <button
        class="ui button"
        data-id="2"
      >
        刪除
      </button>
    </td>
  </tr>
  <tr
    class=""
  >
    <td
      class=""
    >
      3
    </td>
    <td
      class=""
    >
      user3
    </td>
    <td
      class="undefined member-not-lock"
    >
      啟動
    </td>
    <td
      class="undefined member-lock"
    >
      <i
        aria-hidden="true"
        class="lock icon"
      />
      鎖定
    </td>
    <td
      class=""
    >
      2019-06-16 09:36:04
    </td>
    <td
      class=""
    >
      <button
        class="ui button"
      >
        編輯
      </button>
      <button
        class="ui button"
        data-id="3"
      >
        刪除
      </button>
    </td>
  </tr>
  <tr
    class=""
  >
    <td
      class=""
    >
      4
    </td>
    <td
      class=""
    >
      user4
    </td>
    <td
      class="undefined member-not-lock"
    >
      啟動
    </td>
    <td
      class="undefined member-lock"
    >
      <i
        aria-hidden="true"
        class="lock icon"
      />
      鎖定
    </td>
    <td
      class=""
    >
      2019-06-16 09:36:04
    </td>
    <td
      class=""
    >
      <button
        class="ui button"
      >
        編輯
      </button>
      <button
        class="ui button"
        data-id="4"
      >
        刪除
      </button>
    </td>
  </tr>
  <tr
    class=""
  >
    <td
      class=""
    >
      5
    </td>
    <td
      class=""
    >
      user5
    </td>
    <td
      class="undefined member-not-lock"
    >
      啟動
    </td>
    <td
      class="undefined member-lock"
    >
      <i
        aria-hidden="true"
        class="lock icon"
      />
      鎖定
    </td>
    <td
      class=""
    >
      2019-06-16 09:36:04
    </td>
    <td
      class=""
    >
      <button
        class="ui button"
      >
        編輯
      </button>
      <button
        class="ui button"
        data-id="5"
      >
        刪除
      </button>
    </td>
  </tr>
</tbody>
`;
